python/dgl/contrib/dis_kvstore.py

```diff
@@ -220,195 +363,320 @@ def _push_handler(self, name, ID, data):
+                    self._data_store[tensor_name] = F.zerocopy_from_dlpack(dlpack)
+                    self._has_data.add(tensor_name)
+
+    def push(self, name, id_tensor, data_tensor):
+        """Push message to KVServer.
 
         Note that push() is an async operation that will return immediately after calling.
 
         Parameters
         ----------
         name : str
             data name
-        server_id : int
-            target server id
         id_tensor : tensor (mx.ndarray or torch.tensor)
-            a vector storing the ID list
+            a vector storing the global data ID
         data_tensor : tensor (mx.ndarray or torch.tensor)
-            a tensor with the same row size of id
+            a tensor with the same row size of data ID
         """
-        assert server_id >= 0, 'server_id (%d) cannot be a negative number' % server_id
-        assert server_id < self._server_count, 'server_id (%d) must be smaller than server_count' % server_id
+        assert len(name) > 0, 'name cannot be empty.'
         assert F.ndim(id_tensor) == 1, 'ID must be a vector.'
         assert F.shape(id_tensor)[0] == F.shape(data_tensor)[0], 'The data must has the same row size with ID.'
-        msg = KVStoreMsg(
-            type=KVMsgType.PUSH,
-            rank=self._client_id,
-            name=name,
-            id=id_tensor,
-            data=data_tensor)
-        _send_kv_msg(self._sender, msg, server_id)
 
-    def pull(self, name, server_id, id_tensor):
-        """Pull sparse message from KVServer
+        # partition data (we can move this part of code into C-api if needed)
+        server_id = self._data_store[name+'-part-'][id_tensor]
+        # sort index by server id
+        sorted_id = F.tensor(np.argsort(F.asnumpy(server_id)))
+        id_tensor = id_tensor[sorted_id]
+        data_tensor = data_tensor[sorted_id]
+        server, count = np.unique(F.asnumpy(server_id), return_counts=True)
+        # push data to server by order
+        start = 0
+        for idx in range(len(server)):
+            end = start + count[idx]
+            if start == end: # don't have any data for target server
+                continue
+            partial_id = id_tensor[start:end]
+            partial_data = data_tensor[start:end]
+
+            if server[idx] in self._local_server_id:
+                if (name+'-g2l-' in self._has_data) == True:
+                    local_id = self._data_store[name+'-g2l-'][partial_id]
+                else:
+                    local_id = partial_id
+                self._push_handler(name+'-data-', local_id, data_tensor, self._data_store)
+            else:
+                msg = KVStoreMsg(
+                    type=KVMsgType.PUSH, 
+                    rank=self._client_id, 
+                    name=name,
+                    id=partial_id, 
+                    data=partial_data)
+                _send_kv_msg(self._sender, msg, server[idx])
+
+            start += count[idx]
```

Comment:
ideally, this should be implemented in C. we should benchmark this in the real workloads.

Reply:
Sure. We can benchmark the overhead in real distributed GNN tasks. I test it on KGE tasks, and I found that this code does not take too much time.